Graph Support

samuelcolvin opened the pull request:
TODO:
* [x] ~~nodes via decorator~~ impossible without HKT
* [x] infer graph name
* [x] tests
* [x] docs
* [x] examples

This is a work in progress, it's the result of a lot of discussion with @dmontagu.

The idea is to provide a graph/state machine library to use with PydanticAI that is as type-safe as possible in python.

**NOTE: the vast majority of multi-agent examples I've seen to not need a graph or state machine, and would be more complex to write and understand if written using one. You should only use this functionality if:**
1. you understand how to use `Agent` as tools
2. you've tried using standard programming techniques to link agents 
3. after that you're still sure you need a graph library and state machine

In particular this means we define edges (which nodes in a graph can breached from any given node) using type annotations, rather than some separate `set_edges` mechanism.

To do this we define nodes as types (that must inherit from `BaseNode`), to route the graph to (say) `NodeB`, `NodeB` will return an instance of `NodeB` which holders the input data to `NodeB`. Similarly to end a run, nodes should return `End`.

We inspect the return annotation of the `run` method on nodes to build the graph.

here's a minimal simple example:

<details>
<summary>Code</summary>

```py
from __future__ import annotations as _annotations

from dataclasses import dataclass

from pydantic_graph import Graph, BaseNode, End


@dataclass
class NodeA(BaseNode[None]):
    apple: int

    async def run(self, ctx) -> NodeB:
        return NodeB(self.apple / 2)


@dataclass
class NodeB(BaseNode[None]):
    banana: float

    async def run(self, ctx) -> NodeC:
        return NodeC((int(self.banana), int(self.banana) + 4))


@dataclass
class NodeC(BaseNode[None, float]):
    pair: tuple[int, int]

    async def run(self, ctx) -> NodeA | End[float]:
        v1, v2 = self.pair
        if v1 + v2 > 10:
            return End((v1 + v2) / 3)
        else:
            return NodeA(v1 + v2)


graph = Graph(nodes=(NodeA, NodeB, NodeC))
print(graph.mermaid_code(start_node=NodeA))


result, history = graph.run_sync(None, NodeA(8))
print(result)

result, history = graph.run_sync(None, NodeA(7))
print(result)
```
</details>

The mermaid chart printed in the example looks like this:

```mermaid
---
title: graph
---
stateDiagram-v2
  [*] --> NodeA
  NodeA --> NodeB
  NodeB --> NodeC
  NodeC --> NodeA
  NodeC --> [*]
```

And the rest of the output is:

```
4.0
4.666666666666667
```

The graph library is completely independent of LLM use cases, but can relatively easily be used with `pydantic-ai`'s `Agent`, see the `examples/pydantic_ai_examples/email_extract_graph.py` example.

brettkromkamp commented:
In my opinion this feature is critical for adoption of the PydanticAI framework. Any timeframe when this will land in main?

samuelcolvin commented:
@brettkromkamp we'll do our best to get something merged and released this week.

samuelcolvin commented:
I've added support for `Interrupt` to interrupt a run and continue it from the right place.

Here's an example using it:

```py
from __future__ import annotations as _annotations

from dataclasses import dataclass

from typing_extensions import assert_never

from pydantic_ai import Agent
from pydantic_ai.messages import ModelMessage
from pydantic_ai.format_as_xml import format_as_xml
from pydantic_ai_graph import Graph, BaseNode, End, GraphContext, Interrupt, AbstractState

ask_agent = Agent('openai:gpt-4o', result_type=str)


@dataclass
class QuestionState(AbstractState):
    ask_agent_messages: list[ModelMessage] | None = None

    def serialize(self) -> bytes | None:
        raise NotImplementedError('TODO')


@dataclass
class Ask(BaseNode[QuestionState]):
    async def run(self, ctx: GraphContext[QuestionState]) -> Interrupt[Answer]:
        result = await ask_agent.run(
            'Ask a simple question with a single correct answer.', message_history=ctx.state.ask_agent_messages
        )
        if ctx.state.ask_agent_messages is None:
            ctx.state.ask_agent_messages = []
        ctx.state.ask_agent_messages += result.all_messages()
        return Interrupt(Answer(result.data))


@dataclass
class Answer(BaseNode[QuestionState]):
    question: str
    answer: str | None = None

    async def run(self, ctx: GraphContext[QuestionState]) -> Evaluate:
        assert self.answer is not None
        return Evaluate(self.question, self.answer)


@dataclass
class EvaluationResult:
    correct: bool
    comment: str


evaluate_agent = Agent(
    'openai:gpt-4o',
    result_type=EvaluationResult,
    system_prompt='Given a question and answer, evaluate if the answer is correct.',
    result_tool_name='evaluation',
)


@dataclass
class Evaluate(BaseNode[QuestionState]):
    question: str
    answer: str

    async def run(self, ctx: GraphContext[QuestionState]) -> Congratulate | Castigate:
        result = await evaluate_agent.run(format_as_xml({'question': self.question, 'answer': self.answer}))
        if result.data.correct:
            return Congratulate(result.data.comment)
        else:
            return Castigate(result.data.comment)


@dataclass
class Congratulate(BaseNode[QuestionState, None]):
    comment: str

    async def run(self, ctx: GraphContext[QuestionState]) -> End:
        print(f'Correct answer! {self.comment}')
        return End(None)


@dataclass
class Castigate(BaseNode[QuestionState]):
    comment: str

    async def run(self, ctx: GraphContext[QuestionState]) -> Ask:
        print(f'Comment: {self.comment}')
        return Ask()


graph = Graph(nodes=(Ask, Answer, Evaluate, Congratulate, Castigate))

print(graph.mermaid_code(start_node=Ask))
graph.mermaid_save('questions_graph.svg', start_node=Ask)


async def main():
    next_node = Ask()
    state = QuestionState()
    while True:
        result, _ = await graph.run(state, next_node)
        if isinstance(result, End):
            break
        elif isinstance(result, Interrupt):
            assert isinstance(result.node, Answer)
            next_node = result.node
            next_node.answer = input(f'{next_node.question} ')
        else:
            assert_never(result)


if __name__ == '__main__':
    import asyncio

    asyncio.run(main())
```

Which has the following graph:

```mermaid
graph TD
  START --> Ask(Ask)
  Ask(Ask) --> Answer
  Answer[Answer] --> Evaluate
  Evaluate(Evaluate) --> Congratulate
  Evaluate(Evaluate) --> Castigate
  Congratulate(Congratulate) --> END
  Castigate(Castigate) --> Ask
```

You'll see that the `Answer` node is square, not rounded to identify a point where the graph may restart after interruption.

brettkromkamp commented:
Just wondering if the new interrupt mechanism can be used for HITL-purposes? Or, is it more for retrying steps in case of failures. It could also be a general mechanism for all kinds of purposes. I'll take a closer look... definitely exciting to see how this feature is developing, though.

dmontagu commented:
> Just wondering if the new interrupt mechanism can be used for HITL-purposes? Or, is it more for retrying steps in case of failures. It could also be a general mechanism for all kinds of purposes. I'll take a closer look... definitely exciting to see how this feature is developing, though.

It is definitely explicitly and primarily intended for facilitating HITL; if it's useful for other purposes then of course that's great but most of the discussion we've been having about the feature has been oriented around how to use it for human feedback.

dmontagu commented on pydantic_ai_graph/pydantic_ai_graph/state.py:
```diff
@@ -62,10 +84,17 @@ class EndEvent(Generic[StateT, RunEndT]):
-StepOrEnd = Union[Step[StateT, RunEndT], EndEvent[StateT, RunEndT]]
+def _deep_copy_state(state: StateT) -> StateT:
+    if state is None:
+        return None  # pyright: ignore[reportReturnType]
```
changing this to `return state` might eliminate the type error without changing the logic

dmontagu commented on pydantic_ai_graph/pydantic_ai_graph/state.py:
```diff
@@ -25,15 +25,19 @@ def serialize(self) -> bytes | None:
         """Serialize the state object."""
         raise NotImplementedError
```
I wonder if we can eliminate this `AbstractState` by moving the serialization and/or copying logic to be kwargs of the graph, and if not provided, use `copy.deepcopy` (or noop if `None` as you've done) for copying, and `pydantic_core.to_json` for serialization. That would let you use a typical basemodel/dataclass/typeddict as state with minimal boilerplate

dmontagu commented on pydantic_ai_graph/pydantic_ai_graph/nodes.py:
```diff
@@ -69,19 +71,48 @@ def get_node_def(cls, local_ns: dict[str, Any] | None) -> NodeDef[StateT, NodeRu
     next_node_ids: set[str]
+    """IDs of the nodes that can be called next."""
     returns_end: bool
+    """The node definition returns an `End`, hence the node and end the run."""
```
```suggestion
    """The node definition returns an `End`, hence the node can end the run."""
```

dmontagu commented on pydantic_ai_graph/pydantic_ai_graph/state.py:
```diff
@@ -25,15 +25,19 @@ def serialize(self) -> bytes | None:
 @dataclass
-class Step(Generic[StateT, RunEndT]):
-    """History item describing the execution of a step of a graph."""
+class NextNodeEvent(Generic[StateT, RunEndT]):
```
```suggestion
class NodeEvent(Generic[StateT, RunEndT]):
```
?

samuelcolvin commented:
> Just wondering if the new interrupt mechanism can be used for HITL-purposes? Or, is it more for retrying steps in case of failures. It could also be a general mechanism for all kinds of purposes. I'll take a closer look... definitely exciting to see how this feature is developing, though.

@brettkromkamp I've removed `Interrupt` and replaced with what we think is a better API, I've updated my example above.

samuelcolvin commented on pydantic_ai_graph/pydantic_ai_graph/mermaid.py:
```diff
@@ -88,6 +97,8 @@ def node_ids(node_idents: Sequence[NodeIdent] | NodeIdent) -> Iterable[str]:
 class MermaidConfig(TypedDict, total=False):
     """Parameters to configure mermaid chart generation."""
 
+    start_node: Sequence[NodeIdent] | NodeIdent
```
I went back and fourth on this, it felt more natural to use `start_node` and allow it to be multiple, especially as it'll commonly be only one node. But happy to change if you're convinced.

ME-Msc commented:
Hi, team. I have some questions about graph support.

1. It seems that a node's multiple out-edges refer to different transition conditions. Will it be possible to see the condition on the mermaid graph? 
2. Is there any support for run the subsequent nodes in parallel (mentioned in #529 [comments](https://github.com/pydantic/pydantic-ai/issues/529#issuecomment-2558833987) by @izzyacademy ) ？

samuelcolvin commented:
> 1. It seems that a node's multiple out-edges refer to different transition conditions. Will it be possible to see the condition on the mermaid graph?

Hi @ME-Msc, I'm not exactly sure what you mean here? 

I'm going to provide a way to label an edge, but you won't be able to "see the logic" that leads to an edge being followed, as that's just procedural python code.

> 2. Is there any support for running the subsequent nodes in parallel (mentioned in [Functionality to Define Multi-Agent Graphs and Workflows #529](https://github.com/pydantic/pydantic-ai/issues/529) by @izzyacademy ) ？

Not yet, we might add it in future.

izzyacademy commented:
@samuelcolvin I think I have an idea of what he is asking. I had similar thoughts earlier.

It seems @ME-Msc is looking for a mechanism to annotate (within the docstring) the pydantic_graph.BaseNode.run() method with a small note/comment to indicate what condition causes this node to route to the next node returned by this node. It looks like we could parse the docstring for a special tag or something from BaseNode.run() to get a list of conditions and then inject this into the mermaid code generated so that it shows up in the graph image generated.

@ME-Msc I think the goal of the project is to avoid fancy syntax that does not give you visibility into how the parallel nodes/tasks are run. What I would recommend is for you to dedicate a node that can aggregate all the parallel tasks and then spin up async tasks in that node using regular python code that you have 100% visibility and control so that you can see the exceptions, cancellations etc without having to get stressed out when things deviate from the happy path. I hope this helps.

I am working on an example for this because I think many users will have the similar questions/needs based on how they are using other frameworks with custom syntax for routing to parallel nodes in graph transitions

samuelcolvin commented:
@ME-Msc I've updated the example above to:
1. Show an `Edge` annotation
2. Show docstings becoming notes in the mermaid chart

izzyacademy commented on pydantic_graph/pydantic_graph/graph.py:
```diff
@@ -0,0 +1,145 @@
+    async def run(
+        self,
+        state: StateT,
+        node: BaseNode[StateT, RunEndT],
```
We can rename node to start_node in the future to indicate that this is really the beginning of the graph.

izzyacademy replied:
Just a suggestion

izzyacademy replied:
So I think if the parameter is named as start_node it will be clearer (more obvious) to the caller of the run() method that is the start of the graph

samuelcolvin replied:
thanks, fixed.

izzyacademy reviewed (commented):
Thank you for the hard work on implementing this. A lot of folks in the community have been waiting for this.

ME-Msc commented:
@samuelcolvin Thanks for your reply. I'm sure the new example code has solved my first question. What I want is just the annotation of the edge.

```
async def run(self, ctx: GraphContext[QuestionState]) -> Annotated[Evaluate, Edge(label='answer the question')]:
   assert self.answer is not None
   return Evaluate(self.question, self.answer)
```

ME-Msc commented:
@izzyacademy Thank you for your explanation. As for the second question, I am afraid that aggregating all the parallel tasks in one node only by regular Python code (asyncio, multi-threads, etc.) cannot fix it elegantly. There are 3 questions I suggest you consider in the example.

1. What if the parallel tasks are executed by different pydanticAI-agents?
2. How should you manage it if each parallel task has a different number of steps?
3. How do you handle the situation where one step of a parallel task needs to jump to another node(or another step of another node) outside of this node?

I'm looking forward to your example very much!

samuelcolvin commented:
> @izzyacademy Thank you for your explanation. As for the second question, I am afraid that aggregating all the parallel tasks in one node only by regular Python code (asyncio, multi-threads, etc.) cannot fix it elegantly. There are 3 questions I would like to suggest you consider in the example.

We might add support for running multiple nodes at once in future, but I think what we have now is already pretty powerful

> 1. What if the parallel tasks are executed by different pydanticAI-agents?

That should be fine, you can run multiple agents simultaneously just as easily as running the same agent simultaneously.

> 2. How should you manage it if each parallel task has a different number of steps?

You can use all the existing tools in the python toolbox to run multiple tasks in parallel

> 3. How do you handle the situation where one step of a parallel task needs to jump to another node(or another step of another node) outside of this node?

That won't work yet.

izzyacademy commented:
@ME-Msc Thanks for your follow up. @samuelcolvin thanks for responding.

In my personal opinion, it is always best to keep things simple in the design. It makes my life easier as an architect/engineer in production.

While a lot of frameworks may try to give you cool syntax to run nodes in your graph concurrently, in the end what the end up doing is using thread pools to run the cpu-bound tasks or asyncio to run the i/o-bound tasks. The only downside is that you do not have much visibility into what is happening or how that is implemented and in production, that is generally not a good idea.

If you also take a close look at the scenarios, they all gather at some point for every node in the parallel execution to complete or fail before moving on and I don't see hops or jumps in a well designed use case.

I believe that the current mechanism of implementing state transitions with pydantic-graph still allows you to run multiple nodes in parallel even if you have multiple agents running concurrently or only included if certain conditions are true.

So my approach will be to design you graph transitions such that if you need to run multiple "nodes" in parallel (fixed list or variable list) you can compile a dynamic/static list of Coroutines and then gather for them to complete or fail with asycnio and then you decide what to do next from there for the (gathering point node). This also allows you to cancel any task you no longer need with full control and visibility.

If they are cpu-bound you can use threads or other mechanisms to execute these "nodes" within the (gathering point node) and then proceed from there.

My examples will illustrate these. 

I have travel agent scenarios where my agent/s will need to (after my airline or train reservation is booked to a distant city) make inquiry about availability for car rental, hotels, spa and city tours concurrently. 

This variable list of tasks vary depending on the customer profile/preferences.

The customer may only need one or all of the searches. 

All these searches must be done in parallel and then when we find out their results (pass or fail) before we can then proceed to charging the customer payment method for the total price and then sending out a confirmation message.

My point, at this time, is to keep things simple, avoid complex/fancy syntax and have full visibility/control into the inner workings of your code and flow so that you don't have to shave your head while maintaining the app in production :)

I hope this helps.

Thanks for the questions.

I am very excited about this addition to the project and it will definitely make a lot of lives easier when it lands in main and is released to the community.

Thanks Pydantic Team for an amazing effort in the design and implementation of this new capability.

hwong557 commented on docs/graph.md:
```diff
@@ -0,0 +1,359 @@
+
+While this library is developed as part of the PydanticAI; it has no dependency on `pydantic-ai` and can be considered as a pure graph library. You may find it useful whether or not you're using PydanticAI or even building with GenAI.
+
+`pydantic-graph` is designed for advanced users and makes heavy use of Python generics and types hints. It is not designed to b as beginner-friendly as PydanticAI.
```
```suggestion
`pydantic-graph` is designed for advanced users and makes heavy use of Python generics and types hints. It is not designed to be as beginner-friendly as PydanticAI.
```

samuelcolvin replied:
thanks, fixed.

dmontagu commented on Makefile:
```diff
@@ -34,7 +34,7 @@ typecheck-pyright:
 .PHONY: typecheck-mypy
 typecheck-mypy:
-	uv run mypy --strict tests/typed_agent.py
+	uv run mypy
```
Does this cover the `tests/` files? If so, was there a reason we didn't do this sooner?

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+!!! danger "Don't use a nail gun unless you need a nail gun"
+    If PydanticAI [agents](agents.md) are a hammer, and [multi-agent workflows](multi-agent-applications.md) are a sledgehammer, then graphs are a nail gun, with flames down the side:
+
+    * sure, nail guns look cooler than hammers
+    * but nail guns take a lot more setup than hammers
+    * and nail guns don't make you a better builder, they make you a builder with a nail gun
+    * Lastly, (and at the risk of torturing this metaphor), if you're a fan of medieval tools like mallets and untyped Python, you probably won't like nail guns or our approach to graphs. (But then again, if you're not a fan of type hints in Python, you've probably already bounced off PydanticAI to use one of the toy agent frameworks — good luck, and feel free to borrow my sledgehammer when you r ...
```
1, I'm not sure if the metaphors here are sound (do nail guns really take a lot of setup? lol), and 2, I feel like this warning bubble is perhaps a bit overly strongly stated.

I feel like the current phrasing comes across a bit like you are comparing people who are afraid of the graph-based approach to people that who are afraid of types. And I don't think that's the intent, so I think it makes sense to try to just focus on "if you don't think you need this, you probably don't".

samuelcolvin replied:
we can tone it down a bit, but right now I think this is an okay start.

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+    In short, graphs are a powerful tool, but they're not the right tool for every job. Please consider other [multi-agent approaches](multi-agent-applications.md) before proceeding.
+
+    Unless you're sure you need a graph, you probably don't.
```
```suggestion
    If you're not confident a graph-based approach is a good idea, it's probably not necessary.
```
Just trying to soften the language a bit, the current phrasing somehow sounds just lightly condescending to me, if not toward the reader, then toward the library.

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+Alongside PydanticAI, we've developed `pydantic-graph` — an async graph and state machine library for Python where nodes and edges are defined using type hints.
+
+While this library is developed as part of the PydanticAI; it has no dependency on `pydantic-ai` and can be considered as a pure graph library. You may find it useful whether or not you're using PydanticAI or even building with GenAI.
```
```suggestion
While this library has been developed as part of PydanticAI; it has no dependency on `pydantic-ai` and can be considered as a pure graph-based state machine library. You may find it useful whether or not you're using PydanticAI or even building with GenAI.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+While this library is developed as part of the PydanticAI; it has no dependency on `pydantic-ai` and can be considered as a pure graph library. You may find it useful whether or not you're using PydanticAI or even building with GenAI.
+
+`pydantic-graph` is designed for advanced users and makes heavy use of Python generics and types hints. It is not designed to be as beginner-friendly as PydanticAI.
```
> as beginner-friendly as PydanticAI

I hope the usage of generics in PydanticAI can be considered beginner friendly 😅

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+## Graph Types
+
+Graphs are made up of a few key components:
```
I feel like I want to qualify the "Graph" here in some way. The lightest version of that would be putting it in backticks (i.e., "`Graph`s"), but saying something like "Execution Graphs" or "State-machine Graphs" or similar might help here. Just to clarify that this isn't like, a general purpose graph library — it's a state machine library for managing code execution that can be modeled as a graph.

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+### End
+
+[`End`][pydantic_graph.nodes.End] — return value to indicates the graph run should end.
```
```suggestion
[`End`][pydantic_graph.nodes.End] — return value to indicate the graph run should end.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+### GraphContext
+
+[`GraphContext`][pydantic_graph.nodes.GraphContext] — The context for the graph run, similar to PydanticAI's [`RunContext`][pydantic_ai.tools.RunContext], this holds the state of the graph and is passed to nodes when they're run.
```
```suggestion
[`GraphContext`][pydantic_graph.nodes.GraphContext] — The context for the graph run, similar to PydanticAI's [`RunContext`][pydantic_ai.tools.RunContext]. This holds the state of the graph and is passed to nodes when they're run.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+### Nodes
+
+Subclasses of [`BaseNode`][pydantic_graph.nodes.BaseNode] to define nodes.
```
```suggestion
Subclasses of [`BaseNode`][pydantic_graph.nodes.BaseNode] define nodes for execution in the graph.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+Subclasses of [`BaseNode`][pydantic_graph.nodes.BaseNode] to define nodes.
+
+Nodes which are generally [`dataclass`es][dataclasses.dataclass] include:
```
```suggestion
Nodes, which are generally [`dataclass`es][dataclasses.dataclass], include:
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+Nodes which are generally [`dataclass`es][dataclasses.dataclass] include:
+
+* any parameters required when calling the node
```
```suggestion
* fields containing any parameters required when calling the node
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+* any parameters required when calling the node
+* the business logic to execute the node
+* return annotations which are read by `pydantic-graph` to determine the outgoing edges of the node
```
```suggestion
* return annotations, which are read by `pydantic-graph` to determine the outgoing edges of the node
```
(might make sense to say which method's return annotations will be used for this; it strikes me as weird to read "return annotations" on its own in the description of a class.)

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+Nodes are generic in:
+
+* **state** which must have the same type as the state of graphs they're included in, [`StateT`][pydantic_graph.state.StateT] has a default of `None`, so if you're not using state you can omit this generic parameter
```
```suggestion
* **state**, which must have the same type as the state of graphs they're included in, [`StateT`][pydantic_graph.state.StateT]. This TypeVar has a default of `None`, so if you're not using state you can omit this generic parameter
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+Nodes are generic in:
+
+* **state** which must have the same type as the state of graphs they're included in, [`StateT`][pydantic_graph.state.StateT] has a default of `None`, so if you're not using state you can omit this generic parameter
+* **graph return type** this only applies if the node returns [`End`][pydantic_graph.nodes.End], [`RunEndT`][pydantic_graph.nodes.RunEndT] has a default of [Never][typing.Never] so this generic parameter can be omitted if the node doesn't return `End`, but must be included if it does.
```
```suggestion
* **graph return type** — this only applies if the node returns [`End`][pydantic_graph.nodes.End]. [`RunEndT`][pydantic_graph.nodes.RunEndT] has a default of [Never][typing.Never] so this generic parameter can be omitted if the node doesn't return `End`, but must be included if it does.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+### Graph
+
+[`Graph`][pydantic_graph.graph.Graph] — The graph itself, made up of a set of nodes.
```
```suggestion
[`Graph`][pydantic_graph.graph.Graph] — this is the execution graph itself, made up of a set of node classes (i.e., `BaseNode` subclasses).
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+## Stateful Graphs
+
+TODO introduce state
+
+TODO link to issue about persistent state.
```
Not sure if you wanted to update this before merging, but here's a reminder in case you forgot. I'm okay either way.

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+2. A dictionary of products mapped to prices.
+3. The `InsertCoin` node, [`BaseNode`][pydantic_graph.nodes.BaseNode] is parameterized with `MachineState` as that's the state used in this graph.
+4. The `InsertCoin` node prompts the user to insert coins. We keep things simple by just entering a monetary amount as a float. Before you start thinking this is a toy too since it's using [rich's `Prompt.ask`][rich.prompt.PromptBase.ask] within nodes, see [below](#custom-control-flow) for how control flow can be managed when nodes require external input.
+5. The `CoinsInserted` node, again this is a [`dataclass`][dataclasses.dataclass], in this case with one field `amount`, thus nodes calling `CoinsInserted` must provide an amount.
```
```suggestion
5. The `CoinsInserted` node; again this is a [`dataclass`][dataclasses.dataclass], in this case with one field `amount`, thus nodes calling `CoinsInserted` must provide an amount.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+11. If the balance is insufficient, to go `InsertCoin` to prompt the user to insert more coins.
+12. If the product is invalid, go to `SelectProduct` to prompt the user to select a product again.
+13. The graph is created by passing a list of nodes to [`Graph`][pydantic_graph.graph.Graph]. Order of nodes is not important, but will alter how [diagrams](#mermaid-diagrams) are displayed.
+14. Initialize the state, this will be passed to the graph run and mutated as the graph runs.
```
```suggestion
14. Initialize the state. This will be passed to the graph run and mutated as the graph runs.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+12. If the product is invalid, go to `SelectProduct` to prompt the user to select a product again.
+13. The graph is created by passing a list of nodes to [`Graph`][pydantic_graph.graph.Graph]. Order of nodes is not important, but will alter how [diagrams](#mermaid-diagrams) are displayed.
+14. Initialize the state, this will be passed to the graph run and mutated as the graph runs.
+15. Run the graph with the initial state, since the graph can be run from any node, we must pass the start node, in this case `InsertCoin`. [`Graph.run`][pydantic_graph.graph.Graph.run] returns a tuple of the return value (`None`) in this case, and the [history][pydantic_graph.state.HistoryStep] of the graph run.
```
```suggestion
15. Run the graph with the initial state. Since the graph can be run from any node, we must pass the start node — in this case, `InsertCoin`. [`Graph.run`][pydantic_graph.graph.Graph.run] returns a tuple of the return value (`None`) in this case, and the [history][pydantic_graph.state.HistoryStep] of the graph run.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+13. The graph is created by passing a list of nodes to [`Graph`][pydantic_graph.graph.Graph]. Order of nodes is not important, but will alter how [diagrams](#mermaid-diagrams) are displayed.
+14. Initialize the state, this will be passed to the graph run and mutated as the graph runs.
+15. Run the graph with the initial state, since the graph can be run from any node, we must pass the start node, in this case `InsertCoin`. [`Graph.run`][pydantic_graph.graph.Graph.run] returns a tuple of the return value (`None`) in this case, and the [history][pydantic_graph.state.HistoryStep] of the graph run.
+16. The return type of the node's [`run`][pydantic_graph.nodes.BaseNode.run] method is important, it's used to determine the outgoing edges of the node, this in turn is used to render [mermaid diagrams](#mermaid-diagrams) and is enforced at runtime.
```
```suggestion
16. The return type of the node's [`run`][pydantic_graph.nodes.BaseNode.run] method is important, as it is used to determine the outgoing edges of the node. This information is in turn is used to render [mermaid diagrams](#mermaid-diagrams) and is enforced at runtime to detect misbehavior as soon as possible.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+14. Initialize the state, this will be passed to the graph run and mutated as the graph runs.
+15. Run the graph with the initial state, since the graph can be run from any node, we must pass the start node, in this case `InsertCoin`. [`Graph.run`][pydantic_graph.graph.Graph.run] returns a tuple of the return value (`None`) in this case, and the [history][pydantic_graph.state.HistoryStep] of the graph run.
+16. The return type of the node's [`run`][pydantic_graph.nodes.BaseNode.run] method is important, it's used to determine the outgoing edges of the node, this in turn is used to render [mermaid diagrams](#mermaid-diagrams) and is enforced at runtime.
+17. The return type of `CoinsInserted`s [`run`][pydantic_graph.nodes.BaseNode.run] method is a union, meaning multiple outgoing edges are possible.
```
```suggestion
17. The return type of `CoinsInserted`'s [`run`][pydantic_graph.nodes.BaseNode.run] method is a union, meaning multiple outgoing edges are possible.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+15. Run the graph with the initial state, since the graph can be run from any node, we must pass the start node, in this case `InsertCoin`. [`Graph.run`][pydantic_graph.graph.Graph.run] returns a tuple of the return value (`None`) in this case, and the [history][pydantic_graph.state.HistoryStep] of the graph run.
+16. The return type of the node's [`run`][pydantic_graph.nodes.BaseNode.run] method is important, it's used to determine the outgoing edges of the node, this in turn is used to render [mermaid diagrams](#mermaid-diagrams) and is enforced at runtime.
+17. The return type of `CoinsInserted`s [`run`][pydantic_graph.nodes.BaseNode.run] method is a union, meaning multiple outgoing edges are possible.
+18. Unlike other nodes `Purchase` can end the run, so the [`RunEndT`][pydantic_graph.nodes.RunEndT] generic parameter must be set, in this case it's `None` since the graph run return type is `None`.
```
```suggestion
18. Unlike other nodes, `Purchase` can end the run, so the [`RunEndT`][pydantic_graph.nodes.RunEndT] generic parameter must be set. In this case it's `None` since the graph run return type is `None`.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+## Custom Control Flow
+
+In many real-world applications, Graphs cannot run uninterrupted from start to finish — they require external input or run over an extended period of time where a single process cannot execute the entire graph.
```
```suggestion
In many real-world applications, Graphs cannot run uninterrupted from start to finish — they might require external input, or run over an extended period of time such that a single process cannot execute the entire graph run from start to finish without interruption.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+1. Create the state object which will be mutated by [`next`][pydantic_graph.graph.Graph.next].
+2. The start node is `Ask` but will be updated by [`next`][pydantic_graph.graph.Graph.next] as the graph runs.
+3. The history of the graph run is stored in a list of [`HistoryStep`][pydantic_graph.state.HistoryStep] objects, again [`next`][pydantic_graph.graph.Graph.next] will update this list in place.
```
```suggestion
3. The history of the graph run is stored in a list of [`HistoryStep`][pydantic_graph.state.HistoryStep] objects. Again, [`next`][pydantic_graph.graph.Graph.next] will update this list in place.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+
+## Mermaid Diagrams
+
+Pydantic Graph, can generate [mermaid](https://mermaid.js.org/) [`stateDiagram-v2`](https://mermaid.js.org/syntax/stateDiagram.html) diagrams for graphs, as shown above.
```
```suggestion
Pydantic Graph can generate [mermaid](https://mermaid.js.org/) [`stateDiagram-v2`](https://mermaid.js.org/syntax/stateDiagram.html) diagrams for graphs, as shown above.
```

dmontagu commented on docs/graph.md:
```diff
@@ -0,0 +1,714 @@
+* [`Graph.mermaid_image`][pydantic_graph.graph.Graph.mermaid_image] to generate an image of the graph using [mermaid.ink](https://mermaid.ink/)
+* [`Graph.mermaid_save`][pydantic_graph.graph.Graph.mermaid_save] to generate an image of the graph using [mermaid.ink](https://mermaid.ink/) and save it to a file
+
+Beyond the diagrams shown above, you can also customise mermaid diagrams with the following options:
```
```suggestion
Beyond the diagrams shown above, you can also customize mermaid diagrams with the following options:
```

dmontagu commented on docs/graph.md:
this file is impressive

dmontagu commented on docs/multi-agent-applications.md:
````diff
@@ -330,11 +330,6 @@ graph TB
   seat_preference_agent --> END
 ```
 
-## PydanticAI Graphs
-
-!!! example "Work in progress"
-    This is a work in progress and not yet documented, see [#528](https://github.com/pydantic/pydantic-ai/issues/528) and [#539](https://github.com/pydantic/pydantic-ai/issues/539)
-
````
I feel it's worth keeping this section with just a link to the other page, so you don't have to scroll back up to the original numbered list to get the link if you read this far in the docs and are curious about the graph stuff.

dmontagu commented on pydantic_graph/README.md:
```diff
@@ -0,0 +1,16 @@
+Graph and finite state machine library.
+
+This library is developed as part of the [PydanticAI](https://ai.pydantic.dev), however it has no dependency
+on `pydantic-ai` or related packages and can be considered as a pure graph library.
```
```suggestion
on `pydantic-ai` or related packages and can be considered as a pure graph-based state machine library.
```
(I think I made a similar suggestion the first time I encountered this phrasing earlier in the PR.)

dmontagu commented on pydantic_graph/pydantic_graph/nodes.py:
```diff
@@ -0,0 +1,162 @@
+
+
+@dataclass
+class GraphContext(Generic[StateT]):
```
```suggestion
class GraphRunContext(Generic[StateT]):
```
I'm assuming this only applies to graph runs? If so, I think calling it this would make that more clear. Take it or leave it.

dmontagu commented on pydantic_graph/pydantic_graph/nodes.py:
```diff
@@ -0,0 +1,162 @@
+class NodeDef(Generic[StateT, NodeRunEndT]):
+    """Definition of a node.
+
+    This is an internal representation of a node, it shouldn't be necessary to use it directly.
```
```suggestion
    This is an internal representation of a node; it shouldn't be necessary to use it directly.
```

dmontagu replied:
If this is the case, does it need to be a public type?

dmontagu commented on pydantic_graph/pydantic_graph/state.py:
```diff
@@ -0,0 +1,133 @@
+
+    state: StateT
+    """The state of the graph before the node is run."""
+    # node: Annotated[BaseNode[StateT, RunEndT], pydantic.WrapSerializer(node_serializer), pydantic.PlainValidator(node_validator)]
```
```suggestion
```
?

dmontagu commented on pydantic_graph/pydantic_graph/state.py:
```diff
@@ -0,0 +1,133 @@
+    """The duration of the node run in seconds."""
+    kind: Literal['node'] = 'node'
+    """The kind of history step, can be used as a discriminator when deserializing history."""
+    # waiting for https://github.com/pydantic/pydantic/issues/11264, should in InitVar
```
```suggestion
    # waiting for https://github.com/pydantic/pydantic/issues/11264, should be an InitVar
```
?

dmontagu commented on pydantic_graph/pydantic_graph/state.py:
```diff
@@ -0,0 +1,133 @@
+    snapshot_state: Annotated[Callable[[StateT], StateT], pydantic.Field(exclude=True, repr=False)] = field(
+        default=deep_copy_state, repr=False
+    )
+    """Function to snapshot the state of the graph."""
```
I feel this deserves a bit more documentation about why you might want to override the value. In particular, adding a note like "uses copy.deepcopy by default; you may want to provide a different implementation if your state object is memory intensive and largely immutable" (or similar)

samuelcolvin replied:
ye agreed, I think this can come later.

dmontagu commented on tests/graph/test_graph.py:
Nice how self-contained this stuff is; the testability of it is honestly a good argument for reimplementing agent using graph imo

dmontagu reviewed (approved):
Looks good, lots of copy edits and maybe a few meatier comments above, but I think it's ready

izzyacademy commented on pydantic_graph/pydantic_graph/nodes.py:
```diff
@@ -0,0 +1,162 @@
+    """The state of the graph."""
+
+
+class BaseNode(ABC, Generic[StateT, NodeRunEndT]):
```
I wanted to find out if we really need this type NodeRunEndT or if we can just use only RunEndT.

It appears that within the Graph class all the methods are using RunEndT and NodeRunEndT is only used in nodes.py but in graph.py all the references for this type are RunEndT.

It is just my observation so I wanted to confirm if NodeRunEndT is really needed.

Thanks.

samuelcolvin replied:
they have different meanings, so I think it's fine to have different versions.

izzyacademy commented:
Another suggestion before the release of the graph library will be to remove the pydantic-ai-graph and pydantic-graph packages (0.0.1) from the central repositories because the generic types and method signatures in those versions are drastically different and not compatible with what is currently in the git repo for the graph branch

cxbxmxcx commented:
> ~I've added support for `Interrupt` to interrupt a run and continue it from the right place.~
> 
> I've remove `Interrupt` and instead added `next()` method which provides the same functionality, but is more flexible and easier to understand.
> 
> Here's an example using it:
> 
> ```python
> from __future__ import annotations as _annotations
> 
> from dataclasses import dataclass
> from typing import Annotated
> 
> import logfire
> 
> from pydantic_ai import Agent
> from pydantic_ai.messages import ModelMessage
> from pydantic_ai.format_as_xml import format_as_xml
> from pydantic_graph import Graph, BaseNode, End, GraphContext, AbstractState, Edge
> 
> logfire.configure()
> ask_agent = Agent('openai:gpt-4o', result_type=str)
> 
> 
> @dataclass
> class QuestionState(AbstractState):
>     ask_agent_messages: list[ModelMessage] | None = None
> 
>     def serialize(self) -> bytes | None:
>         raise NotImplementedError('TODO')
> 
> 
> @dataclass
> class Ask(BaseNode[QuestionState]):
>     """Generate a question to ask the user.
> 
>     Uses the GPT-4o model to generate a question.
>     """
>     async def run(self, ctx: GraphContext[QuestionState]) -> Annotated[Answer, Edge(label='ask the question')]:
>         result = await ask_agent.run(
>             'Ask a simple question with a single correct answer.', message_history=ctx.state.ask_agent_messages
>         )
>         if ctx.state.ask_agent_messages is None:
>             ctx.state.ask_agent_messages = []
>         ctx.state.ask_agent_messages += result.all_messages()
>         return Answer(result.data)
> 
> 
> @dataclass
> class Answer(BaseNode[QuestionState]):
>     question: str
>     answer: str | None = None
> 
>     async def run(self, ctx: GraphContext[QuestionState]) -> Annotated[Evaluate, Edge(label='answer the question')]:
>         assert self.answer is not None
>         return Evaluate(self.question, self.answer)
> 
> 
> @dataclass
> class EvaluationResult:
>     correct: bool
>     comment: str
> 
> 
> evaluate_agent = Agent(
>     'openai:gpt-4o',
>     result_type=EvaluationResult,
>     system_prompt='Given a question and answer, evaluate if the answer is correct.',
>     result_tool_name='evaluation',
> )
> 
> 
> @dataclass
> class Evaluate(BaseNode[QuestionState]):
>     question: str
>     answer: str
> 
>     async def run(self, ctx: GraphContext[QuestionState]) -> Congratulate | Castigate:
>         result = await evaluate_agent.run(format_as_xml({'question': self.question, 'answer': self.answer}))
>         if result.data.correct:
>             return Congratulate(result.data.comment)
>         else:
>             return Castigate(result.data.comment)
> 
> 
> @dataclass
> class Congratulate(BaseNode[QuestionState, None]):
>     comment: str
> 
>     async def run(self, ctx: GraphContext[QuestionState]) -> End:
>         print(f'Correct answer! {self.comment}')
>         return End(None)
> 
> 
> @dataclass
> class Castigate(BaseNode[QuestionState]):
>     comment: str
> 
>     async def run(self, ctx: GraphContext[QuestionState]) -> Ask:
>         print(f'Comment: {self.comment}')
>         return Ask()
> 
> 
> graph = Graph(nodes=(Ask, Answer, Evaluate, Congratulate, Castigate))
> 
> 
> @dataclass
> class QuestionState(AbstractState):
>     ask_agent_messages: list[ModelMessage] | None = None
> 
>     def serialize(self) -> bytes | None:
>         raise NotImplementedError('TODO')
> 
> 
> print(graph.mermaid_code(start_node=Ask))
> graph.mermaid_save('questions_graph.svg', start_node=Ask)
> 
> 
> async def main():
>     node = Ask()
>     state = QuestionState()
>     history = []
>     with logfire.span('run questions graph'):
>         while True:
>             node = await graph.next(state, node, history)
>             if isinstance(node, End):
>                 print('\n'.join(e.summary() for e in history))
>                 break
>             elif isinstance(node, Answer):
>                 node.answer = input(f'{node.question} ')
>             # otherwise just continue
> 
> 
> if __name__ == '__main__':
>     import asyncio
> 
>     asyncio.run(main())
> ```
> 
> Which has the following graph:
> 
> ```mermaid
> stateDiagram-v2
>   [*] --> Ask
>   Ask --> Answer: ask the question
>   note right of Ask
>     Generate a question to ask the user.
>     Uses the GPT-4o model to generate a question.
>   end note
>   Answer --> Evaluate: answer the question
>   Evaluate --> Congratulate
>   Evaluate --> Castigate
>   Congratulate --> [*]
>   Castigate --> Ask
> ```
> 
> ~You'll see that the `Answer` node is square, not rounded to identify a point where the graph may restart after interruption.~ this is no longer included, I don't think that matters.

First off just want to say this looks and runs great. Thanks for all the effort on this. I wrote a book called AI Agents In Action (Manning) and used Behavior Trees to control agents but I also really like this approach.

Question: I pulled the latest pydantic-graph from PyPi (could not find a repo for this) and the Edge class is missing. I quickly had Claude generate me one so I have things working. Is there a repo for this work?

cxbxmxcx commented:
> @cxbxmxcx the work is still in progress. You can clone the repo and then switch to the **graph** branch
> 
> From there you can install hatch and hatchling and then use hatch to build the .whl file for local installation and tests
> 
> ```shell
> git clone git@github.com:pydantic/pydantic-ai.git 
> cd pydantic-ai
> 
> git checkout graph
> 
> cd pydantic_graph
> hatch build
> ```
> 
> I hope this helps.

Thanks for help, yes it should.

c3-ali commented on docs/examples/question-graph.md:
````diff
@@ -0,0 +1,37 @@
+
+The mermaid diagram generated in this example looks like this:
+
+```mermaid
````
@samuelcolvin This does not represent the `question_graph.py` anymore. Could you update and use the output of `mermaid_image` instead?

c3-ali replied:
fyi @dmontagu
